feat: add HTTP queue metrics for NIM frontend request tracking

keivenchang opened the pull request:
#### Overview:

This PR implements HTTP queue metrics for frontend request tracking to monitor request processing pipeline performance and detect engine pool exhaustion.

#### Details:

- Add new HTTP queue gauge metric to track requests waiting in HTTP processing queue
- Implement HttpQueueGuard RAII object for automatic queue time tracking from HTTP handler start to metrics processing
- Integrate HTTP queue tracking in OpenAI chat completions handler
- Add HTTP_QUEUE prometheus metric name constant
- Track time from request arrival until engine processing begins to identify bottlenecks

#### Where should the reviewer start?

- `lib/llm/src/http/service/metrics.rs` - New HTTP queue gauge and HttpQueueGuard implementation
- `lib/llm/src/http/service/openai.rs` - Integration of HTTP queue tracking in chat completions
- `lib/runtime/src/metrics/prometheus_names.rs` - New metric name constant

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

Relates to DIS-564, DIS-563

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Added an HTTP queue metric to monitor the number of requests waiting before processing, exposed via Prometheus with per-model labeling.
  - Improved timing accuracy for request lifecycle metrics.
  - Instrumentation updates in chat completions ensure queue time is measured until processing begins, including streaming and non-streaming paths.

- Documentation
  - Clarified the definition of the inflight requests metric to indicate it covers requests sent to the engine (e.g., vLLM, SGLang).

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

keivenchang committed:
feat: add HTTP queue metrics for frontend request tracking

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
fix: refactor http_queue timing and rename processing functions

- Move http_queue_guard creation to beginning of request handlers
- Move inflight_guard creation after engine.generate() call
- Rename processing functions for clarity (process_streaming_with_*)

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
refactor: move metrics functions and fix error counting

- Move InflightGuard creation before engine.generate() calls for proper error counting
- Move EventConverter and related functions from openai.rs to metrics.rs
- Rename functions:
  - process_streaming_with_metrics -> process_response_and_observe_metrics
  - process_response_using_streaming_event_converter -> process_response_using_event_converter_and_observe_metrics
- Consolidate get_parsing_options function in metrics.rs
- Update grpc/service/openai.rs to use shared metrics functions
- Add documentation for HTTP queue vs inflight metrics

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang commented on lib/llm/src/http/service/openai.rs:
```diff
@@ -457,6 +457,11 @@ async fn chat_completions(
+    let http_queue_guard = state
+        .metrics_clone()
+        .create_http_queue_guard(&request.inner.model);
+
```
Next PR.

keivenchang committed:
fix: move http_queue_guard creation after template resolution

Ensures consistent model labels across all metrics for the same request.

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
refactor: move metrics parsing options and rename HTTP queue metric

- Move get_parsing_options from grpc service to http metrics module
- Rename HTTP_QUEUE to HTTP_QUEUED_REQUESTS for clarity
- Update comment to specify 'first response' timing

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

nnshah1 commented on lib/llm/src/http/service/metrics.rs:
```diff
@@ -18,13 +25,77 @@ use super::RouteDoc;
     inter_token_latency: HistogramVec,
 }
 
+/*
```
@keivenchang can we add this to metrics readme?

nnshah1 commented on lib/llm/src/http/service/metrics.rs:
```diff
@@ -18,13 +25,77 @@ use super::RouteDoc;
+- 7 requests will be in HTTP queue most of the time
+- All 10 requests will be counted as inflight (from start until complete response)
+
+Testing Setup:
```
maybe the testing setup could be moved to a test? - for future

nnshah1 commented on lib/llm/src/http/service/metrics.rs:
```diff
@@ -279,6 +370,13 @@ impl Metrics {
     /// The [`InflightGuard`] is an RAII object will handle incrementing the inflight gauge and
     /// request counters.
+    ///
+    /// # Metrics Distinction
```
we should note that queue time will also include prefill time -

keivenchang replied:
Yes, done.

nnshah1 commented on lib/llm/src/http/service/metrics.rs:
```diff
@@ -467,6 +596,117 @@ impl Drop for ResponseMetricCollector {
+///
+/// Creates parsing options with tool call parser and reasoning parser for the specified model.
+/// Currently reasoning parser is not implemented (returns None).
+pub fn get_parsing_options(
```
seems a little odd to have this in metrics - @grahamking , @keivenchang is there another place this would fit naturally? in a way to be shared?

keivenchang replied:
Since this is related to ModelManager, I moved it to model_manager.rs (PR coming up).

nnshah1 reviewed (approved):
LGTM - a couple thoughts:

1) is there another spot for the response functions? maybe response_handler.rs - having them in metrics.rs seems a little unintuitive.

2) could we make the names shorter? (not sure what would work here) but maybe a single method with flags for the behavior? not sure that makes sense.

nnshah1 commented on lib/runtime/src/metrics/prometheus_names.rs:
```diff
@@ -45,7 +45,10 @@ pub mod frontend_service {
 
-    /// Number of inflight requests
+    /// Number of requests waiting in HTTP queue before receiving the first response.
+    pub const HTTP_QUEUED_REQUESTS: &str = "http_queued_requests";
```
can we make this "queued_requests" instead of "http_queued_requests" - not sure the "http" part is adding value here

keivenchang committed:
fix: resolve merge conflict and update metrics naming

- Resolved merge conflict in lib/llm/src/http/service/metrics.rs by combining client_disconnect_gauge and http_queue_gauge fields
- Updated metric names to use consistent naming convention (added _total suffix)
- Moved detailed metrics documentation to deploy/metrics/README.md
- Added comprehensive documentation for HTTP queue vs inflight metrics distinction
- Updated prometheus_names.rs with centralized metric name constants

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
refactor: move get_parsing_options from metrics.rs to ModelManager

- Move get_parsing_options function from lib/llm/src/http/service/metrics.rs to lib/llm/src/discovery/model_manager.rs as a method
- Update all call sites in openai.rs and kserve.rs to use the new ModelManager method
- Remove the old function from metrics.rs
- Improves code organization by placing parsing options functionality with ModelManager where it belongs
- All tests pass, functionality remains unchanged

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
docs: add Prometheus metric naming conventions to prometheus_names.rs

- Add comprehensive naming convention guidelines with prefix/name/suffix structure
- Include common transformations from vague suffixes to specific ones
- Provide clear examples of correct vs incorrect naming patterns
- Document avoidance of vague suffixes like _counter, _gauge, _time, _size, _rate
- Add visual indicators (❌/✅) for easy reference
- Ensure consistency with Prometheus best practices

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>
